### PR TITLE
Fix Migrations Stability

### PR DESCRIPTION
**Pull request by PineappleIOnic**
## What does this PR do?

(Provide a description of what this PR does and why it's needed.)

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Screenshots may also be helpful.)

## Related PRs and Issues

- (Related PR or issue)

## Checklist

- [ ] Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
- [ ] If the PR includes a change to an API's metadata (desc, label, params, etc.), does it also include updated API specs and example docs?

### CODE REVIEW

File: app/controllers/api/migrations.php [hunk 1]
```diff
@@ -309,7 +310,7 @@
     ->label('sdk.response.type', Response::CONTENT_TYPE_JSON)
     ->label('sdk.response.model', Response::MODEL_MIGRATION)
     ->param('resources', [], new ArrayList(new WhiteList(NHost::getSupportedResources())), 'List of resources to migrate')
-    ->param('subdomain', '', new URL(), 'Source\'s Subdomain')
+    ->param('subdomain', '', new Text(512), 'Source\'s Subdomain')
```
Comment (eldadfux) [inline app/controllers/api/migrations.php:313]: Can we use a host or domain validator instead? Why just a simple text, what do we expect here?

File: src/Appwrite/Auth/OAuth2/Firebase.php [hunk 2]
```diff
@@ -226,32 +251,92 @@ public function assignIAMRoles(string $accessToken, string $email, string $proje
         $this->request('POST', 'https://cloudresourcemanager.googleapis.com/v1/projects/' . $projectId . ':setIamPolicy', [
             'Authorization: Bearer ' . \urlencode($accessToken),
             'Content-Type: application/json'
-        ], json_encode([
+        ], \json_encode([
             'policy' => $iamRoles
         ]));
     }
 
+    function generateRandomString($length = 10)
```
Comment (eldadfux) [inline src/Appwrite/Auth/OAuth2/Firebase.php:259]: Missing encapsulation.

File: src/Appwrite/Auth/OAuth2/Firebase.php [hunk 3]
```diff
@@ -226,32 +251,92 @@ public function assignIAMRoles(string $accessToken, string $email, string $proje
...
+        $characters = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
+        $charactersLength = strlen($characters);
+        $randomString = '';
+        for ($i = 0; $i < $length; $i++) {
+            $randomString .= $characters[random_int(0, $charactersLength - 1)];
+        }
+        return $randomString;
+    }
+
+    public function createCustomRole(string $accessToken, string $projectId): array
```
Comment (eldadfux) [inline src/Appwrite/Auth/OAuth2/Firebase.php:270]: This has to be public?

File: app/init.php [hunk 4]
```diff
@@ -135,6 +135,7 @@
 const APP_SOCIAL_DEV = 'https://dev.to/appwrite';
 const APP_SOCIAL_STACKSHARE = 'https://stackshare.io/appwrite';
 const APP_SOCIAL_YOUTUBE = 'https://www.youtube.com/c/appwrite?sub_confirmation=1';
+const APP_INTERNAL_LOOPBACK_HOST = 'appwrite';
```
Comment (eldadfux) [inline app/init.php:138]:
```suggestion
const APP_HOSTNAME_INTERNAL = 'appwrite';
```

**Review by eldadfux — CHANGES_REQUESTED** [review]

File: src/Appwrite/Auth/OAuth2/Firebase.php (hunk 3 above)
Comment (PineappleIOnic) [reply src/Appwrite/Auth/OAuth2/Firebase.php]: No it doesn't mb, changing to private

**Review by eldadfux — APPROVED** [review]

**Pull request merged**